src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,804 @@
+	g_autofree char *url = NULL;
+	const char *remote_name;
+	GError *internal_error = NULL;
+	GsApp *ext_runtime;
+	GsPluginData *priv;
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	priv = gs_plugin_get_data (plugin);
+
+	ext_runtime = gs_plugin_get_app_external_runtime (plugin, app);
+
+	if (!ext_runtime) {
+		g_debug ("External app '%s' didn't have any asset! "
+			 "Not installing and marking as state unknown!",
+			 gs_app_get_id (app));
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+
+		return TRUE;
+	}
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
+		if (!build_runtime (plugin, app, ext_runtime, error)) {
+			g_debug ("Failed to build runtime '%s'", runtime);
+			return FALSE;
+		}
+
+		if (!gs_flatpak_refine_app (priv->flatpak, ext_runtime,
+					    GS_PLUGIN_REFINE_FLAGS_DEFAULT,
+					    cancellable, error)) {
+			g_debug ("Failed to refine '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+	}
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
+		g_debug ("Cannot install app asset '%s': state is unknown",
+			 gs_app_get_id (ext_runtime));
+		return FALSE;
+	}
+
+	switch (gs_app_get_state (ext_runtime)) {
+	case AS_APP_STATE_INSTALLED:
+		g_debug ("App asset '%s' is already installed",
+			 gs_app_get_id (ext_runtime));
+		break;
+
+	case AS_APP_STATE_UPDATABLE:
+		g_debug ("Updating '%s'", gs_app_get_id (ext_runtime));
+
+		if (!gs_flatpak_update_app (priv->flatpak, ext_runtime,
+					    cancellable, error)) {
+			g_debug ("Failed to update '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+		break;
+
+	case AS_APP_STATE_AVAILABLE:
+		g_debug ("Installing '%s'", gs_app_get_id (ext_runtime));
+		gs_flatpak_app_install (priv->flatpak, ext_runtime, cancellable,
+					error);
+		if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
+			g_debug ("Failed to install '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+
+		/* Clean-up remote (we only needed it for installing the app) */
+		remote_name = gs_app_get_origin (ext_runtime);
+		if (!flatpak_remote_delete (remote_name,
+					    &internal_error)) {
+			if (internal_error) {
+				g_debug ("Failed to delete remote '%s': %s",
+					 remote_name, internal_error->message);
+				g_error_free (internal_error);
+			}
+		}
+
+		break;
+
+	case AS_APP_STATE_UNKNOWN:
+	default:
+		/* In case we may end up here somehow, just let the situation
+		 * be dealt with in the check for the 'installed' state below */
+		break;
+	}
+
+	if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
+		gs_app_set_state (app, gs_app_get_state (ext_runtime));
+		g_debug ("Cannot install '%s' because runtime '%s' is not "
+			 "installed", gs_app_get_id (app),
+			 gs_app_get_id (ext_runtime));
+		return FALSE;
```

Comment:
In a lot of these places I don't set errors because the upstream code also doesn't do it but I agree it may seem inconsistent so I can set them.